core/src/subgraph/instance_manager.rs

```diff
@@ -317,3 +258,490 @@ impl EventConsumer<SubgraphAssignmentProviderEvent> for SubgraphInstanceManager
+                    "id" => id_for_err.to_string()
+                );
+            }
+            CancelableError::Error(e) => {
+                error!(
+                    logger_for_err,
+                    "Subgraph instance failed to run: {}", e;
+                    "id" => id_for_err.to_string()
+                );
+
+                // Set subgraph status to Failed
+                let status_ops =
+                    SubgraphDeploymentEntity::update_failed_operations(&id_for_err, true);
+                if let Err(e) = store_for_err.apply_entity_operations(status_ops, EventSource::None)
+                {
+                    error!(
+                        logger_for_err,
+                        "Failed to set subgraph status to Failed: {}", e;
+                        "id" => id_for_err.to_string()
+                    );
+                }
+            }
+        })
+}
+
+fn process_block<B, S, T>(
+    logger: Logger,
+    ctx: IndexingContext<B, S, T>,
+    block_stream_cancel_handle: CancelHandle,
+    needs_restart: Arc<AtomicBool>,
+    block: EthereumBlock,
+) -> impl Future<Item = IndexingContext<B, S, T>, Error = CancelableError<Error>>
+where
+    B: BlockStreamBuilder,
+    S: ChainStore + Store,
+    T: RuntimeHostBuilder,
+{
+    let logger = logger.new(o!(
+        "block_number" => format!("{:?}", block.block.number.unwrap()),
+        "block_hash" => format!("{:?}", block.block.hash.unwrap())
+    ));
+
+    // Extract logs relevant to the subgraph
+    let instance = ctx.state.instance.clone();
+    let logs: Vec<_> = block
+        .transaction_receipts
+        .iter()
+        .flat_map(|receipt| {
+            receipt
+                .logs
+                .iter()
+                .filter(|log| instance.read().unwrap().matches_log(&log))
+        })
+        .cloned()
+        .collect();
+
+    if logs.len() == 0 {
+        debug!(logger, "No events found in this block for this subgraph");
+    } else if logs.len() == 1 {
+        info!(logger, "1 event found in this block for this subgraph");
+    } else {
+        info!(
+            logger,
+            "{} events found in this block for this subgraph",
+            logs.len()
+        );
+    }
+
+    // Obtain current and new block pointer (after this block is processed)
+    let block = Arc::new(block);
+    let block_ptr_now = EthereumBlockPointer::to_parent(&block);
+    let block_ptr_after = EthereumBlockPointer::from(&*block);
+    let block_ptr_for_new_data_sources = block_ptr_after.clone();
+
+    // Clone a few things to pass into futures
+    let logger1 = logger.clone();
+    let logger2 = logger.clone();
+    let logger3 = logger.clone();
+    let logger4 = logger.clone();
+
+    // Process events one after the other, passing in entity operations
+    // collected previously to every new event being processed
+    process_logs(
+        logger.clone(),
+        ctx.state.instance.clone(),
```

Comment:
afaict this is the one place that makes us need the `Arc<RwLock` around the instance. If we instead keep the pattern we use everywhere else by having `process_logs` take the ctx by value and then return it, I believe we can mutate the instance without needing synchronization.

Reply:
Aaaaand done 😀

Reply:
awesome